rayon-core/src/sleep/counters.rs

```diff
@@ -0,0 +1,225 @@
+use std::sync::atomic::{AtomicU64, Ordering};
+
+pub(super) struct AtomicCounters {
+    /// Packs together a number of counters.
+    ///
+    /// Consider the value `0x1111_2222_3333_4444`:
+    ///
+    /// - The bits 0x1111 are the **jobs counter**.
+    /// - The bits 0x2222 are the **sleepy counter**.
+    /// - The bits 0x3333 are the number of **idle threads**.
+    /// - The bits 0x4444 are the number of **sleeping threads**.
+    ///
+    /// See the struct `Counters` below.
+    value: AtomicU64,
```

Comment:
As we noted in chat, `AtomicU64` is only available starting in Rust 1.34 (which is close to a year old for compat), but even then it's not necessarily available on all architectures. We should probably stick to `AtomicUsize`, though we need to consider what that means for 32-bit.

AIUI the jobs/sleepy counters don't really care about true values, just comparing relative change, right? Will it be a problem if that has only an 8-bit rollover on 32-bit targets? Then the idle/sleeping threads seem like hard limits on our thread count, which we should enforce somewhere.